src/sdam/monitor.rs

```diff
@@ -56,84 +86,109 @@ pub(super) fn monitor_server(
-    conn: &mut Connection,
-    mut server_type: ServerType,
-    server: &Weak<Server>,
-) -> Option<ServerType> {
-    // If the server has been dropped, terminate the monitoring thread.
-    let server = match server.upgrade() {
-        Some(server) => server,
-        None => return None,
-    };
-
-    // If the topology has been dropped, terminate the monitoring thread.
-    let topology = match server.topology() {
-        Some(topology) => topology,
-        None => return None,
-    };
-
-    // Send an isMaster to the server.
-    let server_description = check_server(conn, server_type, &server);
-    server_type = server_description.server_type;
-
-    update_topology(topology, server_description);
-
-    Some(server_type)
-}
+    /// Checks the the server by running an `isMaster` command. If an I/O error occurs, the
+    /// connection will replaced with a new one.
+    fn check_server_and_update_topology(&mut self) {
+        // If the server has been dropped, terminate the monitoring thread.
+        let server = match self.server.upgrade() {
+            Some(server) => server,
+            None => return,
+        };
+
+        // If the topology has been dropped, terminate the monitoring thread.
+        let topology = match server.topology() {
+            Some(topology) => topology,
+            None => return,
+        };
+
+        // Send an isMaster to the server.
+        let server_description = self.check_server();
+        self.server_type = server_description.server_type;
+
+        update_topology(topology, server_description);
+    }
 
-fn check_server(
-    conn: &mut Connection,
-    server_type: ServerType,
-    server: &Arc<Server>,
-) -> ServerDescription {
-    let address = conn.address().clone();
+    fn check_server(&mut self) -> ServerDescription {
+        let address = self.address.clone();
 
-    match is_master(conn) {
-        Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
-        Err(e) => {
-            server.clear_connection_pool();
+        match self.is_master() {
+            Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
+            Err(e) => {
+                self.clear_connection_pool();
 
-            if server_type == ServerType::Unknown {
-                return ServerDescription::new(address, Some(Err(e)));
+                if self.server_type == ServerType::Unknown {
+                    return ServerDescription::new(address, Some(Err(e)));
+                }
             }
         }
+
+        ServerDescription::new(address, Some(self.is_master()))
     }
 
-    ServerDescription::new(address, Some(is_master(conn)))
-}
+    fn is_master(&mut self) -> Result<IsMasterReply> {
+        let result = self.is_master_inner();
+
+        if result
+            .as_ref()
+            .err()
+            .map(|e| e.kind.is_network_error())
+            .unwrap_or(false)
+        {
+            self.connection = self.make_new_connection()?;
```

Comment:
To consolidate the existing threads, I'll just start a new one here.
This connection re-establishment logic makes a lot more sense to me if it is included with the rest of the error handling done in `check_server`, and seems out of place here inside of the `is_master` helper itself.
e.g. I'm thinking our `check_server` could look very similar to the pseudocode presented in the SDAM spec:
```
def checkServer(description):
    try:
        call ismaster
        return new ServerDescription
    except NetworkError as e0:
        clear connection pool for the server

        if description.type is Unknown or PossiblePrimary:
            # Failed on first try to reach this server, give up.
            return new ServerDescription with type=Unknown, error=e0
        else:
            # We've been connected to this server in the past, retry once.
            try:
                reconnect and call ismaster
                return new ServerDescription
            except NetworkError as e1:
                return new ServerDescription with type=Unknown, error=e1
```

Reply:
It's somewhat tough to model the code like that given that Rust has no concept of nested try/catch; the closest analog to try/catch we have is `?`, which always returns from the function on an error. We'd have to break it up into multiple functions, which is what I've done here.

Reply:
That makes sense; I see why it's necessary as is. I like the name change to `perform_is_master`, only remaining suggestion I have wrt this is to update `check_server` to remove the `return` from the success branch, and move the retry attempt into the error branch. I think that better communicates the actual control flow, and it more closely mirrors the spec as well.

Reply:
Ah, yeah, that does make it much more similar; done